### ISSUE TITLE
Flow to TypeScript, TypeScript to Flow

### ISSUE THREAD

**Issue by goodmind** [esmbly/design#1]
How would this work if they are really different type systems?

**Comment by ollelauribostrom** [comment]
@goodmind Not sure if I understand your question 🙂On a high level the transformation between Flow and TypeScript (and the other way around) will work like this:

- The TypeScript file is parsed to an AST
- Each node in the AST is processed
- Base on the type of node, either perform some transformation (using [recast](https://github.com/benjamn/recast#usage)) - or just return the node as-is (Flow and TypeScript do have some identical syntax)
- After each node has been processed, turn the AST back into code and write it to a file

**Comment by goodmind** [comment]
@ollelauribostrom yeah, but there is some syntax that can't be converted without losing type information. Or inferred types that doesn't exist in AST

**Comment by MaxGraey** [comment]
It probably could intermediate AST with greatest common factor for both type system. So more complex entities always disassembly to simpler set of entities and assemble back to higher order in the end of transform if this possible

**Comment by ollelauribostrom** [comment]
@goodmind Sorry for the late reply. You're right, figuring out a way to deal with such cases are definitely on the roadmap for an MVP, do you have any thoughts on how to handle this? 

@MaxGraey suggestion sounds like a good approach. It might also be an option to let the user manually edit any edge case syntax before continuing with the transformation (sort of like: `git merge` --> `fix conflicts..` --> `git merge --continue`). 

Currently, work is happening in the [main repo](https://github.com/esmbly/esmbly) to implement a prototype that is able to transform some really simple programs  🙂

**Comment by MaxGraey** [comment]
btw it may be makes sense using something like [jscodeshift](https://github.com/facebook/jscodeshift) for that instead pure recast package
